File: week-1/3-credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int everyOtherDigit(long creditCard);
int multiplySum(int lastDigit);
int numOfDigits(long creditCard);
bool validAmex(long creditCard, int numDigits);
bool validMaster(long creditCard, int numDigits);
bool visacheck(long creditCard, int numDigits);

int main(void)
{
    long creditCard = get_long("Credit card: ");
    int sumEveryOtherDigit = everyOtherDigit(creditCard);
    int numDigits = numOfDigits(creditCard);
    bool amex = validAmex(creditCard, numDigits);
    bool master = validMaster(creditCard, numDigits);
    bool visa = visacheck(creditCard, numDigits);
    if (sumEveryOtherDigit % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    else if (amex == true)
    {
        printf("AMEX\n");
    }
    else if (master == true)
    {
        printf("MASTERCARD\n");
    }
    else if (visa == true)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

bool validAmex(long creditCard, int numDigits)
{
    int firstTwoDigits = creditCard / pow(10,13);
    if ((numDigits == 15) && (firstTwoDigits == 34 || firstTwoDigits == 37))
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool validMaster(long creditCard, int numDigits)
{
    int firstTwoDigits = creditCard / pow(10, 14);
    if ((numDigits == 16) && (firstTwoDigits > 50 && firstTwoDigits < 56))
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool visacheck(long creditCard, int numDigits)
{
     if (numDigits == 13 || numDigits == 16)
     {
          int i = numDigits - 1;
          int firstnum = creditCard / pow(10, i);

          if (firstnum == 4)
          {
               return true;
          }
          else
          {
               return false;
          }
     }
     return false;
}

int numOfDigits(long creditCard)
{
    int count = 0;
    while(creditCard > 0)
    {
        count = count + 1;
        creditCard = creditCard / 10;
    }
    return count;
}

int everyOtherDigit(long creditCard)
{
    int sum = 0;
    bool isAlternateDigit = false;
    while (creditCard > 0)
    {
        if (isAlternateDigit == true)
        {
            int lastDigit = creditCard % 10;
            int prod = multiplySum(lastDigit);
            sum = sum + prod;
        }
        else
        {
            int lastDigit = creditCard % 10;
            sum = sum + lastDigit;
        }
        isAlternateDigit = !isAlternateDigit;
        creditCard = creditCard / 10;
    }
    return sum;
}

int multiplySum(int lastDigit)
{
    int multiply = lastDigit * 2;
    int sum = 0;
    while (multiply > 0)
    {
        int lastDigitMultiplied = multiply % 10;
        sum = sum +  lastDigitMultiplied;
        multiply = multiply / 10;
    }
    return sum;
}

File: week-6/3-sentimental-credit/credit.py
import math

def main():
    while True:
        try:
            creditCard = int(input("What's your credit card?: "))
            break
        except ValueError:
            print("Input only numbers!")
            print()

    sumOfEveryOtherDigit = everyOtherDigit(creditCard)
    cardDigits = len(str(creditCard))
    validAmex = isAmex(creditCard, cardDigits)
    validMaster = isMaster(creditCard, cardDigits)
    validVisa = isVisa(creditCard, cardDigits)

    if sumOfEveryOtherDigit % 10 != 0:
        print("INVALID")
    elif validAmex == True:
        print("AMEX")
    elif validMaster == True:
        print("MASTERCARD")
    elif validVisa == True:
        print("VISA")
    else:
        print("INVALID")

def everyOtherDigit(creditCard):
    counterSum = 0
    isAlternateDigit = False

    while creditCard > 0:
        if isAlternateDigit == True:
            lastDigit = creditCard % 10
            product = multiplySum(lastDigit)
            counterSum += product
        else:
            lastDigit = creditCard % 10
            counterSum += lastDigit
        isAlternateDigit = not isAlternateDigit
        creditCard = math.floor(creditCard / 10)
    return counterSum


def multiplySum(lastDigit):
    counterSum = 0
    multiply = lastDigit * 2
    while multiply > 0:
        lastDigitMultiplied = multiply % 10
        counterSum += lastDigitMultiplied
        multiply = math.floor(multiply / 10)
    return counterSum

def isAmex(creditCard, cardDigits):
    americEx = math.floor(creditCard / pow(10, 13))
    if cardDigits == 15 and (americEx == 34 or americEx == 37):
        return True
    else:
        return False

def isMaster(creditCard, cardDigits):
    masterCard = math.floor(creditCard / pow(10, 14))
    if cardDigits == 16 and (masterCard >= 51 and masterCard <= 55):
        return True
    else:
        return False

def isVisa(creditCard, cardDigits):
    if cardDigits == 13 or cardDigits == 16:
        power = cardDigits - 1
        visa = math.floor(creditCard / pow(10, power))
        if visa == 4:
            return True
        else:
            return False

if __name__ == "__main__":
    main()
